File: bindings/nodejs/examples/how_tos/advanced_transactions/claim_transaction.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { OutputsToClaim, Wallet, initLogger } from '@iota/sdk';
require('dotenv').config({ path: '.env' });

// Run with command:
// yarn run-example ./how_tos/advanced_transactions/claim_transaction.ts

// This example claims all claimable outputs in the wallet.
async function run() {
    initLogger();
    try {
        for (const envVar of [
            'WALLET_DB_PATH',
            'STRONGHOLD_PASSWORD',
            'EXPLORER_URL',
        ]) {
            if (!(envVar in process.env)) {
                throw new Error(
                    `.env ${envVar} is undefined, see .env.example`,
                );
            }
        }

        const wallet = await Wallet.create({
            storagePath: process.env.WALLET_DB_PATH,
        });

        await wallet.sync();

        // To sign a transaction we need to unlock stronghold.
        await wallet.setStrongholdPassword(
            process.env.STRONGHOLD_PASSWORD as string,
        );

        // Get all claimable outputs
        const output_ids = await wallet.claimableOutputs(OutputsToClaim.All);
        console.log(`Available outputs to claim:`);
        for (const output_id of output_ids) {
            console.log(output_id);
        }

        const transaction = await wallet.claimOutputs(output_ids);
        console.log(`Transaction sent: ${transaction.transactionId}`);

        const blockId = await wallet.reissueTransactionUntilIncluded(
            transaction.transactionId,
        );
        console.log(`Block sent: ${process.env.EXPLORER_URL}/block/${blockId}`);
    } catch (error) {
        console.error('Error: ', error);
    }
}

void run().then(() => process.exit());

File: bindings/python/examples/how_tos/advanced_transactions/claim_transaction.py
import os

from dotenv import load_dotenv
from iota_sdk import Wallet, WalletOptions

# This example uses secrets in environment variables for simplicity which
# should not be done in production.
load_dotenv()

# In this example we will claim outputs that have additional unlock
# conditions as expiration or storage deposit return.

for env_var in ['WALLET_DB_PATH', 'STRONGHOLD_PASSWORD']:
    if env_var not in os.environ:
        raise Exception(f'.env {env_var} is undefined, see .env.example')

wallet = Wallet(WalletOptions(storage_path=os.environ.get('WALLET_DB_PATH')))

wallet.set_stronghold_password(os.environ["STRONGHOLD_PASSWORD"])

# Sync wallet with the node
wallet.sync()

# Only the unspent outputs in the wallet
output_ids = wallet.claimable_outputs('All')

print('Available outputs to claim:')
for output_id in output_ids:
    print(f'{output_id}')

transaction = wallet.claim_outputs(output_ids)
print(f'Transaction sent: {transaction.transaction_id}')

wallet.wait_for_transaction_acceptance(
    transaction.transaction_id)
print(
    f'Tx accepted: {os.environ["EXPLORER_URL"]}/transactions/{transaction.transaction_id}')
